Add sensors for Aprilaire integration

chamberlain2007 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add sensor entities to the Aprilaire integration.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/31773

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

chamberlain2007 committed:
Add sensors for Aprilaire integration

chamberlain2007 committed:
Exclude from coverage

chamberlain2007 committed:
Exclude from coverage

chamberlain2007 committed:
Add comment

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+FAN_STATUS_MAP = {0: "Off", 1: "On"}
+
+
+@dataclass(frozen=True)
+class AprilaireSensorDescriptionMixin:
+    """Mixin for Aprilaire sensor."""
+
+    status_key: str | None
+    status_sensor_available_value: int | None
+    status_sensor_exists_values: list[int] | None
+    value_key: str
+    value_fn: Callable[[Any, Any], Any] | None
+
+
+@dataclass(frozen=True)
+class AprilaireSensorDescription(
+    SensorEntityDescription, AprilaireSensorDescriptionMixin
+):
+    """Class describing Aprilaire sensor entities."""
```
```suggestion
@dataclass(frozen=True, kw_only=True)
class AprilaireSensorDescription(SensorEntityDescription):
    """Class describing Aprilaire sensor entities."""

    status_key: str | None
    status_sensor_available_value: int | None
    status_sensor_exists_values: list[int] | None
    value_key: str
    value_fn: Callable[[Any, Any], Any] | None
```

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+    AprilaireSensorDescription(
+        key="humidification_status",
+        translation_key="humidification_status",
+        device_class=None,
+        state_class=None,
```
Can be omitted

chamberlain2007 replied:
Done

joostlek replied:
For some reason they are back

chamberlain2007 replied:
Fixed again, thanks

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+        self,
+        coordinator: AprilaireCoordinator,
+        description: AprilaireSensorDescription,
+        unique_id: str | None,
```
```suggestion
        unique_id: str,
```
You probably did this because the config entry unique id can technically be None. Please assign it to a variable and assert that variable

chamberlain2007 replied:
Done

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+        super().__init__(coordinator, unique_id)
+
+    @property
+    def available(self) -> bool:
```
Since you use a coordinator, that means the BaseAprilaireEntity is inheriting a CoordinatorEntity -> you should also call `super().available`

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+            return True
+
+        return (
+            super().available
```
Oh I see you do now, but can this statement be simplified?

chamberlain2007 replied:
@joostlek I have simplified what I can. Hopefully it is a bit clearer

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,263 @@
+        """Return the value reported by the sensor."""
+        raw_value = self.coordinator.data.get(self.entity_description.value_key)
+
+        return (
+            self.entity_description.value_fn(self, raw_value)
+            if self.entity_description.value_fn
+            else raw_value
+        )
```
Please avoid using inline ternaries if they span more than 1 line

chamberlain2007 replied:
Done

joostlek commented on homeassistant/components/aprilaire/strings.json:
```diff
@@ -23,6 +23,35 @@
+    },
+    "sensor": {
+      "indoor_humidity_controlling_sensor": {
+        "name": "Indoor Humidity Controlling Sensor"
```
Please use sentence case instead of title case

chamberlain2007 replied:
Done

chamberlain2007 committed:
Update homeassistant/components/aprilaire/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

chamberlain2007 committed:
Code review updates

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,254 @@
+        translation_key="humidification_status",
+        device_class=None,
+        state_class=None,
+        status_key=Attribute.HUMIDIFICATION_AVAILABLE,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[1, 2],
+        value_key=Attribute.HUMIDIFICATION_STATUS,
+        value_fn=lambda _, value: HUMIDIFICATION_STATUS_MAP.get(value),
+    ),
+    AprilaireSensorDescription(
+        key="ventilation_status",
+        translation_key="ventilation_status",
+        status_key=Attribute.VENTILATION_AVAILABLE,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[1],
+        value_key=Attribute.VENTILATION_STATUS,
+        value_fn=lambda _, value: VENTILATION_STATUS_MAP.get(value),
+    ),
+    AprilaireSensorDescription(
+        key="air_cleaning_status",
+        translation_key="air_cleaning_status",
+        status_key=Attribute.AIR_CLEANING_AVAILABLE,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[1],
+        value_key=Attribute.AIR_CLEANING_STATUS,
+        value_fn=lambda _, value: AIR_CLEANING_STATUS_MAP.get(value),
+    ),
+    AprilaireSensorDescription(
+        key="fan_status",
+        translation_key="fan_status",
+        status_key=None,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[],
+        value_key=Attribute.FAN_STATUS,
+        value_fn=lambda _, value: FAN_STATUS_MAP.get(value),
```
these status sensors look to be returning a default set of names. You can make these an enum sensor. The added value of this is that you have to pass in a list of possible values (via `options`) and when users want to automate with it for example, they can see what possible values the entity returns.

This does require that the values returned by the sensor are in `snake_case` and then you can add state translations in the strings.json.

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,254 @@
+        AprilaireSensor(coordinator, description, config_entry.unique_id)
+        for description in SENSOR_TYPES
+        if description.status_key is None
+        or description.status_sensor_exists_values is None
+        or coordinator.data.get(description.status_key)
+        in description.status_sensor_exists_values
```
What do we do here?

chamberlain2007 replied:
@joostlek There's not really a better way of writing this I don't think. It is getting all of the possible sensor types and including them if either

1. `status_key` or `status_sensor_exists_values` is `None`, indicating that the sensor should always be added
2. The coordinator data for the `status_key` is one of the allowed values in `status_sensor_exists_values`

The reason for the slightly clunky logic with the `status_sensor_exists_values` is that the thermostat returns a value for the status for which multiple may correspond to the sensor being available. For instance, the `Indoor temperature controlling sensor` should show in the UI if the `INDOOR_TEMPERATURE_CONTROLLING_SENSOR_STATUS` value is either 0 (normal) or 1/2 (installed but in an error state).

joostlek commented on homeassistant/components/aprilaire/sensor.py:
```diff
@@ -0,0 +1,254 @@
+            == self.entity_description.status_sensor_available_value
+        )
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement."""
+        if (
+            self.entity_description.native_unit_of_measurement is None
+            and self.entity_description.state_class is not None
+        ):
+            if self._sensor_option_unit_of_measurement is UNDEFINED:
+                return self.hass.config.units.temperature_unit
+
+            return self._sensor_option_unit_of_measurement
+
+        return super().native_unit_of_measurement
```
What does the Aprilaire device return? In what unit? Is this dynamic?

Looking at the util.py the device always returns celcius. Can you maybe give an example why the special rounding is needed?

chamberlain2007 replied:
The device does always return celsius. The problem is when converting to Fahrenheit, the device itself uses away-from-zero midpoint rounding, whereas Python uses banker's rounding. This leads to inconsistent rounding in the case of values when the user is displaying values in Fahrenheit in HA.

Consider 22.5C which is 72.5F. With away-from-zero rounding, the device displays this as 73F. However with banker's rounding, HA would show this as 72F.

The solution as implemented is to have the sensor always report that it has the same native unit of measurement as it is configured to display in, and then handle the correct rounding when returning the native value.

chamberlain2007 replied:
This actually reminds me that I need to do the same for the climate entity. This fix was made in the HACS repo after I started getting this merged into core, so I never made the change on the core side.

chamberlain2007 replied:
Please let me know if you have any other questions about this

bdraco committed:
Improve performance of image streams (#112810)

jbouwh committed:
Allow Just-in-Time platform setup for mqtt (#112720)

* Allow Just-in-Time platform setup for mqtt

* Only forward the setup of new platforms

* Fix new  platforms being setup at reload + test

* Revert not related changes

* Remove unused partial

* Address comments, only import plaforms if needed

* Apply suggestions from code review

* Add multipl platform discovery test

* Improve test

* Use a lock per platform

catsmanac committed:
Enphase Envoy refactor and extend diagnostics (#109080)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Avoid saving auth right after we load it during startup (#112008)

bdraco committed:
Schedule coroutines with eager_task from async_track_state_change_event (#112807)

joostlek committed:
Remove entity description mixin in IPP (#112781)

jbouwh committed:
Cleanup mqtt PLATFORMS constant and CI-test platform filters (#112847)

bdraco committed:
Only read cpu once during systemmonitor setup (#112863)

* Only read cpu once during systemmonitor setup

* type

bdraco committed:
Fix MJPEG fallback when still image URL is missing with basic auth (#112861)

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

* Fix MJPEG fallback when still image URL is missing with basic auth

I picked up an old DCS-930L (circa 2010) camera to test with
to fix #94877

bdraco committed:
Schedule coroutine functions eagerly when async_listen uses run_immediately (#112846)

We have a few places where we call async_listen with a callback so we can schedule
the coro eagerly. We can drop these in favor of setting run_immediately now.

cdce8p committed:
Update pytest to 8.1.1 (#112859)

bdraco committed:
Refactor handling of device updates in ESPHome (#112864)

Kane610 committed:
Split out deCONZ hub (#112854)

* Rename DeconzGateway to DeconzHub

* Move gateway.py to hub/hub.py

abjorck committed:
Bump yalexs to 2.0.0 (#111706)

Co-authored-by: J. Nick Koston <nick@koston.org>

Ernst79 committed:
Add Locks to Xiaomi-BLE (#111156)

joostlek committed:
Remove entity description mixin in Freebox (#112766)

clssn committed:
Update the numato-gpio dependency of the numato integration to v0.12.0 (#112272)

* Update the numato-gpio dependency of the numato integration to v0.12.0

* Augment numato integration manifest with integration_type

Fulfills a requirement in the PR checklist.

joostlek committed:
Remove entity description mixin in Private BLE Device (#112922)

joostlek committed:
Remove entity description mixin in Netgear (#112911)

joostlek committed:
Remove entity description mixin in SFR Box (#112937)

joostlek committed:
Remove entity description mixin in Renault (#112925)

* Remove entity description mixin in Renault

* Update homeassistant/components/renault/sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* Update homeassistant/components/renault/button.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

---------

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

joostlek committed:
Remove entity description mixin in Roku (#112931)

bachya committed:
Streamline Notion config entry updates (refresh token and user ID) (#112832)

joostlek committed:
Remove entity description mixin in Overkiz (#112914)

joostlek committed:
Remove entity description mixin in Trafikverket Weatherstation (#112957)

joostlek committed:
Remove entity description mixin in Trafikverket Train (#112956)

joostlek committed:
Remove entity description mixin in Trafikverket Ferry (#112955)

joostlek committed:
Remove entity description mixin in Trafikverket Camera (#112954)

joostlek committed:
Remove entity description mixin in Sensibo (#112936)

shaiu committed:
Improve 17track tests (#112734)

* 17Track tests

* add 17Track sensor to coverage

* extract repeated code

* 1. _goto_future - call tick only once
2. change test name to reflect test
3. remove ifs from test

* remove undersocre from _goto_future

jbouwh committed:
Fix mqtt platform setup race (#112888)

bdraco committed:
Run debouncer tasks eagerly to avoid scheduling on the event loop (#112789)

bdraco committed:
Import custom components in the executor by default (#112177)

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

joostlek committed:
Remove entity description mixin in Vodafone Station (#112967)

joostlek committed:
Remove entity description mixin in Zeversolar (#112977)

joostlek committed:
Remove entity description mixin in Youtube (#112975)

joostlek committed:
Remove entity description mixin in WAQI (#112970)

joostlek committed:
Remove entity description mixin in Tradfri (#112953)

gjohansson-ST committed:
Use async_at_started in Speedtest.Net (#112110)

joostlek committed:
Remove entity description mixin in Permobil (#112918)

joostlek committed:
Remove entity description mixin in PEGELONLINE (#112917)

joostlek committed:
Remove entity description mixin in Melcloud (#112905)

joostlek committed:
Remove entity description mixin in Justnimbus (#112893)

joostlek committed:
Remove entity description mixin in LaCrosse View (#112900)

joostlek committed:
Move wsdot fixture to integration test (#112996)

joostlek committed:
Move uk_transport fixture to integration test (#112995)

joostlek committed:
Move OpenALPR cloud fixture to integration test (#112994)

jbouwh committed:
Remove not needed hass.async_block_till_done in test (#113002)

joostlek committed:
Remove entity description mixin in Switcher Kis (#112945)

dknowles2 committed:
Add missing translation placeholder in Hydrawise (#113007)

Add missing translation placeholder

joostlek committed:
Remove entity description mixin in Nettigo Air Monitor (#112909)

joostlek committed:
Remove entity description mixin in Tractive (#112952)

Remove entity description mixin in
Tractive

joostlek committed:
Remove entity description mixin in Synology DSM (#112946)

joostlek committed:
Add icon translations to Picnic (#112181)

* Add icon translations to Picnic

* fix

joostlek committed:
Add icon translations to Husqvarna automower (#111727)

* Add icon translations to Husqvarna automower

* fix

Thomas55555 committed:
Bump aioautomower to 2024.3.0 (#112627)

Fix error in Husqvarna automower in Zones dataclass

joostlek committed:
Remove entity description mixin in Shelly (#112938)

* Remove entity description mixin in Shelly

* fix

* Fix

yuxincs committed:
Migrate APCUPSD to has entity name  (#112997)

* Properly set entity names for APCUPSD

* Add test cases to prevent future regressions

* Fix tests due to the updated entity IDs

* Prettify code

* Remove redundant translation key

dmulcahey committed:
Fix race condition when ZHA group members change (#113030)

bdraco committed:
Ignore logging events in zha websocket tests (#113031)

allenporter committed:
Bump ical to 7.0.1 and always use home assistant timezone for local todo dtstart (#113034)

joostlek committed:
Fix Bang & Olufsen enum naming (#113022)

Kane610 committed:
Bump axis to v53 (#113019)

bdraco committed:
Bump aioesphomeapi to 23.1.1 (#113016)

changelog: https://github.com/esphome/aioesphomeapi/compare/v23.1.0...v23.1.1

Fixes cryptic error message the user tried to send a command
or execute a service call when the device was disconnected

joostlek committed:
Enable more SIM ruff rules (#113015)

* SIM101 SIM103

* SIM107 SIM109

* SIM110

* SIM112 SIM113

* SIM115

* SIM116

* Fix

* Fix

* Fix

bdraco committed:
Migrate duckdns to use async_run_hass_job (#113012)

The code would create a hassjob and than run the wrapped
function with async_run_job so it had to work out the
job type twice

bdraco committed:
Avoid scheduling person updates on the event loop (#113010)

These update call the storage collection async_update_item
which never suspends so they can finish synchronously without
having to be scheduled on the loop when run_immediately,
which schedules the task eagerly is set

Ernst79 committed:
Bump bthome-ble to 3.8.0 (#113008)

Bump bthome-ble

joostlek committed:
Move Microsoft face fixture to integration test (#112993)

* Move Microsoft face fixture to integration test

* Update tests/components/microsoft_face/test_init.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

* Fix

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

joostlek committed:
Move metoffice fixture to integration test (#112992)

* Move metoffice fixture to integration test

* Fix

* Update tests/components/metoffice/test_config_flow.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

* Fix

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

joostlek committed:
Move london_air fixture to integration test (#112991)

* Move london_air fixture to integration test

* Fix

joostlek committed:
Move Feedreader fixture to integration test (#112989)

bdraco committed:
Schedule tasks eagerly when called from hass.add_job (#113014)

joostlek committed:
Move openhardwaremonitor fixture to integration tests (#112982)

* Move openhardwaremonitor fixture

* Update tests/components/openhardwaremonitor/test_sensor.py

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

joostlek committed:
Remove entity description mixin in V2C (#112960)

yuxincs committed:
Remove a redundant check in APCUPSD's config flow (#113032)

Remove a redundant check that is impossible to happen in practice

jbouwh committed:
Rename mqtt platform setup method (#113042)

* Rename mqtt platform setup method

* Format method to one line

bdraco committed:
Fix google assistant cloud sync test (#113044)

bdraco committed:
Fix CoreState in cloud google assistant tests (#113045)

chemelli74 committed:
Allow Shelly CoAP to honour default network adapter (#110997)

* Allow Shelly CoAP to honor default network adapter

* apply review comment

* 1 more debug log line

* adapt code to library changes

* test

* improve test

* one more test

joostlek committed:
Remove entity description mixin in Vilfo (#112966)

joostlek committed:
Remove entity description mixin in Tado (#112947)

Bre77 committed:
Add tests for Aussie Broadband diagnostics (#113049)

* Add tests for diag

* Fix docstring

* Spelling

marcelveldt committed:
Bump aioslimproto to 3.0.0 (#113046)

joostlek committed:
Remove entity description mixin in Opower (#112913)

dependabot[bot] committed:
Bump Wandalen/wretry.action from 1.4.7 to 1.4.8 (#113038)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

FlyingFeng2021 committed:
Bump boschshcpy to 0.2.82 (#112890)

joostlek committed:
Remove entity description mixin in WeatherFlow (#112971)

joostlek committed:
Remove entity description mixin in Starlink (#112943)

joostlek committed:
Remove entity description mixin in Schlage (#112934)

joostlek committed:
Remove entity description mixin in Kaleidescape (#112896)

joostlek committed:
Remove entity description mixin in Launch Library (#112902)

joostlek committed:
Remove entity description mixin in Kaiterra (#112895)

joostlek committed:
Move yandex transport fixture to integration test (#112988)

* Move yandex transport fixture to integration test

* Update tests/components/yandex_transport/test_sensor.py

* Ran ruff

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>
Co-authored-by: Franck Nijhof <git@frenck.dev>

joostlek committed:
Remove entity description mixin in Meater (#112904)

joostlek committed:
Remove entity description mixin in Minecraft Server (#112908)

joostlek committed:
Remove YAML configuration from WAQI (#113027)

autinerd committed:
Bump webmin-xmlrpc to 0.0.2 (#113004)

joostlek committed:
Remove entity description mixin in QNAP QSW (#112924)

joostlek committed:
Remove entity description mixin in Wallbox (#112969)

joostlek committed:
Remove entity description mixin in SABnzbd (#112933)

joostlek committed:
Remove entity description mixin in Honeywell Lyric (#112903)

joostlek committed:
Remove entity description mixin in Z-Wave.js (#112979)

joostlek committed:
Remove entity description mixin in Zamg (#112976)

joostlek committed:
Remove entity description mixin in Yalexs BLE (#112974)

joostlek committed:
Remove entity description mixin in Xiaomi Miio (#112973)

joostlek committed:
Remove entity description mixin in Whirlpool (#112972)

joostlek committed:
Remove entity description mixin in VeSync (#112965)

joostlek committed:
Remove entity description mixin in Venstar (#112964)

joostlek committed:
Remove entity description mixin in Vallox (#112963)

joostlek committed:
Remove entity description mixin in Tuya (#112959)

joostlek committed:
Remove entity description mixin in Transmission (#112958)

joostlek committed:
Remove entity description mixin in Tolo (#112951)

joostlek committed:
Remove entity description mixin in Tautulli (#112949)

joostlek committed:
Remove entity description mixin in Steamist (#112944)

joostlek committed:
Remove entity description mixin in Solaredge (#112942)

joostlek committed:
Remove entity description mixin in Smappee (#112941)

joostlek committed:
Remove entity description mixin in SleepIQ (#112940)

joostlek committed:
Remove entity description mixin in Skybell (#112939)

joostlek committed:
Remove entity description mixin in Roomba (#112932)

joostlek committed:
Remove entity description mixin in Ring (#112928)

joostlek committed:
Remove entity description mixin in Roborock (#112930)

joostlek committed:
Remove entity description mixin in Meteo-France (#112907)

joostlek committed:
Remove entity description mixin in Rituals Perfume Genie (#112929)

joostlek committed:
Remove entity description mixin in Renson (#112926)

joostlek committed:
Remove entity description mixin in Pure Energie (#112923)

joostlek committed:
Remove entity description in Point (#112921)

joostlek committed:
Remove entity description mixin in Picnic (#112920)

joostlek committed:
Remove entity description mixin in Pi-hole (#112919)

joostlek committed:
Remove entity description mixin in PECO (#112915)

joostlek committed:
Remove entity description mixin in Onvif (#112912)

joostlek committed:
Remove entity description mixin in Netatmo (#112910)

joostlek committed:
Remove entity description mixin in Melnor (#112906)

joostlek committed:
Remove entity description mixin in Landisgyr Heat Meter (#112901)

joostlek committed:
Remove entity description mixin in Kraken (#112899)

joostlek committed:
Remove entity description mixin in Kostal Plenticore (#112898)

joostlek committed:
Remove entity description mixin in Juicenet (#112892)

edenhaus committed:
Use UV instead of PIP in the CI (#113051)

marcelveldt committed:
Fix adding Hue bridge manually by IP (#113055)

Ernst79 committed:
Bump xiaomi-ble to 0.27.0 (#113013)

shaiu committed:
Add ConfigFlow for seventeentrack integration (#111196)

* Add config flow to 17Track

* Import config from configuration.yaml

* 1. move import to async_setup_platform
2. add USERNAME (email) in title for uniqueness

* Add options flow

* Add tests

* Add CONF_SHOW_ARCHIVED  and CONF_SHOW_DELIVERED to data from options

* Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/manifest.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

* Update homeassistant/components/seventeentrack/__init__.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

* 1. Added repair issues
2. _async_validate_input inlined
3. added unique id
4. take default scan interval

* fix

* 1. move async_create_issue to async_setup_platform
2. fix tests
3. black + pylint

* combine USER_SCHEMA and OPTIONS_SCHEMA

* small fix

* remove async_setup

* fix tests and add 100% coverage

* 1. remove CONFIG_SCHEMA
2. remove error log
3. add issue with more description when import issues happen
4. some linting

* Update homeassistant/components/seventeentrack/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* use freezer
use AsyncMock
fix tests

* add test_flow_fails
parametrize tests where needed
test_import_flow_already_configured - where a unique id already configured (abort flow)

* lint

* fix rebase issues

* some more fix

* 17Track revert tests and put them in a different PR

* adapt tests to MockConfigEntry

* Update tests/components/seventeentrack/test_sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/seventeentrack/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* 1. create fixture for config and another with options
2. set options with default values
3. remove CONFIG_SCHEMA

* Update tests/components/seventeentrack/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update tests/components/seventeentrack/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* 1. get options from import data and default if not present
2. rename mock_config_entry_no_options -> mock_config_entry_with_default_options

* move ACCOUNT_ID to mock_seventeentrack_api.return_value.profile.account_id

* Apply suggestions from code review

* Update tests/components/seventeentrack/test_config_flow.py

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Christopher Fenner <9592452+CFenner@users.noreply.github.com>

joostlek committed:
Move HomematicIP Cloud fixture to integration test (#112987)

* Move HomematicIP Cloud fixture to integration test

* Update tests/components/homematicip_cloud/helper.py

* Update tests/components/homematicip_cloud/helper.py

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

sdb9696 committed:
Prepare ring update service for deprecation (#108781)

* Prepare ring update service for deprecation

* Update service removal release number

joostlek committed:
Enable even more SIM ruff rules (#113017)

* SIM202 SIM211 SIM220 SIM221 SIM222 SIM223

* SIM910 SIM911

* SIM

* Update homeassistant/components/mqtt/siren.py

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

* Update homeassistant/components/mqtt/siren.py

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

---------

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

marcelveldt committed:
Fix for controlling Hue switch entities (#113064)

folke committed:
Fix hvac_mode for viessmann devices with heatingCooling mode (#113054)

marcelveldt committed:
Fix for Hue not applying effect in scene (#113057)

bdraco committed:
Fix here_travel_time creating many refresh requests at startup (#113041)

joostlek committed:
Remove entity description mixin in Github (#112771)

bdraco committed:
Ensure wemo discovery is run in a background task (#112665)

joostlek committed:
Remove fixtures from global fixtures folder (#113060)

Remove fixtures

joostlek committed:
Improve lists in integrations [A] (#113006)

* Use list comprehension [A]

* Use list comprehension [A]

* Update homeassistant/components/aws/notify.py

OndraM committed:
Change airthings pressure sensor device_class to atmospheric pressure (#113005)

LukasdeBoer committed:
Bump rova to 0.4.1 (#113066)

bachya committed:
Bump `regenmaschine` to 2024.03.0 (#113074)

joostlek committed:
Improve lists in integrations [C-D] (#113072)

alistair23 committed:
components/gardena_bluetooth: Improve avaliability reliability (#113056)

* components/gardena_bluetooth: Improve avaliability reliability

The available() function incorrectly returns false even though the device
is accessible.

The last_update_success property should correctly indicate if the device
isn't contactable, so we don't need to call async_address_present().
This is confirmed by seeing that no other users are calling
async_address_present() in the available() function.

This commit removes the call to async_address_present() to help fix the
sensor being unavailable when using a ESPHome BLE proxy.

Signed-off-by: Alistair Francis <alistair@alistair23.me>

---------

Signed-off-by: Alistair Francis <alistair@alistair23.me>
Co-authored-by: Joakim Plate <elupus@ecce.se>

joostlek committed:
Improve lists in integrations [B] (#113069)

joostlek committed:
Add icon translations to Version (#112343)

marcelveldt committed:
Fix colormode attribute on grouped Hue light (#113071)

balloob committed:
Only load services.yaml for integrations that have it (#112732)

Co-authored-by: J. Nick Koston <nick@koston.org>

RunC0deRun committed:
Fix optional Jellyfin RunTimeTicks (#108254)

bdraco committed:
Start tasks eagerly in for async_at_start(ed) (#112802)

bdraco committed:
Migrate dispatcher to create tasks eagerly (#112845)

There were quite a few of these that never had to suspend.

bdraco committed:
Migrate _async_when_setup to use eager_start (#112872)

This one does not make much difference, but its a lot
easier to do it now instead of in the future

sdb9696 committed:
Ensure that the ring integration always raises HomeAssistantError for user actions (#109893)

* Wrap library exceptions in HomeAssistantErrors

* Remove commented

* Update post review

* Update post second review

bdraco committed:
Fix race in script stop that could cause async_stop to hang forever (#113089)

bdraco committed:
Fix failing google diagnostics test (#113095)

Bre77 committed:
Fix location data in Teslemetry (#112866)

Fix location data in coordinator

mikewoudenberg committed:
Update Loqed helper for more logging (#112646)

Updates Loqed helper for more logging

bdraco committed:
Migrate async_run_job to use eager_start for tasks (#113011)

bdraco committed:
Bump aiodhcpwatcher to 0.8.1 (#113096)

bdraco committed:
Reduce latency to populate initial HKC bluetooth device state (#113103)

Kane610 committed:
Bump axis to v54 (#113091)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Make sonos ssdp discovery callback a callback function (#113107)

bdraco committed:
Make sonos household_coordinator setup a callback function (#113108)

bdraco committed:
Make sonos binary sensor entity creation callback functions (#113109)

bdraco committed:
Add support for run_immediately to async_listen_once (#113020)

bdraco committed:
Run coroutines as eager tasks in async_run_hass_job (#111683)

* Run coroutines as eager tasks in async_run_hass_job

Note that this does not change async_add_hass_job

Do not merge this. For test run only

* Phase out periodic tasks

* false by default or some tests will block forever, will need to fix each one manually

* kwarg works

* kwarg works

* kwarg works

* fixes

* fix more tests

* fix more tests

* fix lifx

* opensky

* pvpc_hourly_pricing

* adjust more

* adjust more

* smarttub

* adjust more

* adjust more

* adjust more

* adjust more

* adjust

* no eager executor

* zha

* qnap_qsw

* fix more

* fix fix

* docs

* its a wrapper now

* add more coverage

* coverage

* cover all combos

* more fixes

* more fixes

* more fixes

* remaining issues are legit bugs in tests

* make tplink test more predictable

* more fixes

* feedreader

* grind out some more

* make test race safe

* limit first scope to triggers

* one more

* Start tasks eagerly in for async_at_start(ed)

A few of these can avoid being scheduled on the loop
during startup

* fix cloud

* Revert "fix cloud"

This reverts commit 5eb3ce695da788bcae649f82c9527c0f9307139c.

* fix test to do what start does

* flip flag

* flip flag

* Fix here_travel_time creating many refresh requests at startup

- Each entity would try to refresh the coordinator which
  created many tasks. Move the refresh to a single
  async_at_started

- The tests fired the EVENT_HOMEASSISTANT_START event
  but the code used async_at_started which only worked
  because the tests did not set CoreState to not_running

* fix azure

* remove kw

* remove kw

* rip

* cover

* more rips

* more rips

* more rips

bdraco committed:
Make cast async_cast_discovered a callback function (#113111)

Nothing was being awaited here and this function
is never subclassed

bdraco committed:
Migrate homekit to use async_at_started (#113102)

bdraco committed:
Make restore state started a callback function (#113110)

bdraco committed:
Use run_immediately for starting discovery at the started event (#113112)

bdraco committed:
Update timezone in supervisor in an eager task (#113113)

bdraco committed:
Migrate scene platform to use async_add_executor_job (#113114)

bdraco committed:
Attach template triggers at start eagerly (#113120)

This method calls async_initialize_triggers which is likely to never
suspend and the attach can avoid being scheduled on the event loop

bdraco committed:
Add run_immediately to the zone core config update async_listen (#113119)

Calling async_update_config will never suspend but cannot be changed to a callback
function because it would break the collections api

bdraco committed:
Remove unnecessary use of async_run_job in script helper (#113118)

The function being passed to `async_run` was always a callback

bdraco committed:
Remove unnecessary use of async_add_job in openalpr_cloud (#113116)

bdraco committed:
Await discovery_function in the DiscoveryFlowHandler instead of wrapping it in a task (#113115)

Await discovery_function in the DiscoveryFlowHandler instead of wrapping

This function was always a coro so it can be awaited
directly instead of wrapping it in add_job

bdraco committed:
Simplify automation startup logic (#113122)

bdraco committed:
Migrate remaining call in check_config helper to use async_get_component (#113123)

bdraco committed:
Import homeassistant trigger platforms in the executor (#113124)

bdraco committed:
Switch the reload helper to use async_get_component (#113126)

definitio committed:
Add A6 family code for DS2438 1-Wire sensor (#112844)

* Add A6 family code for DS2438

* Add tests

* Fix switch

* Apply code review suggestion

* Add comments

Thomas55555 committed:
Add missing device info to Husqvarna Automower (#113090)

* Add missing DeviceInfo to Husqvarna Automower

* add a test

* Adress review

* Update homeassistant/components/husqvarna_automower/entity.py

* fix url

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

slyoldfox committed:
Add scheduled mode to renault charge mode (#105967)

Add scheduled to renault charge mode - fixes #105751

emontnemery committed:
Add reconfigure flow to homeworks (#112419)

* Add reconfigure flow to homeworks

* Fix tests

* Use async_update_reload_and_abort helper

* Try to fix test shutdown

zweckj committed:
bump pytedee_async to 0.2.16 (#113135)

jbouwh committed:
Fix google_asssistant sensor state reporting (#112838)

* Fix post google_assistant sensor values as float not string

* Fix aqi reporting and improve tests

* Fix _air_quality_description_for_aqi and  test

jbouwh committed:
Validate state_class with last_reset_value_template for mqtt sensors (#113099)

zweckj committed:
Include pytedee_async logger in tedee integration (#112590)

add pytedee logger

yozik04 committed:
Update vallox_websocket_api to 5.1.1 (#113139)

Bump vallox-websocket-api==5.1.1

dependabot[bot] committed:
Bump Wandalen/wretry.action from 1.4.8 to 1.4.9 (#113132)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bdraco committed:
Convert sets to list before passing to the JSON serializer in the registries (#113133)

There were a few places we were missing the set to list
conversions in the registries. We do this before its
cached to avoid the JSON serializer having to fallback
to the default method every time since its expensive
to switch back from the native code into python context
for every set.

bdraco committed:
Add the ability to create Debouncer tasks as background tasks (#113128)

* Add the ability to Debouncer tasks in the background

This is a more general solution as a followup to
https://github.com/home-assistant/core/pull/112652#discussion_r1517159607

* Add the ability to Debouncer tasks in the background

This is a more general solution as a followup to
https://github.com/home-assistant/core/pull/112652#discussion_r1517159607

* fix

synesthesiam committed:
Add device_id to sentence trigger and external conversation APIs (#113094)

* Add device_id to sentence trigger and external conversation APIs

* Remove device_id from external API

* Update tests/components/conversation/snapshots/test_init.ambr

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

joostlek committed:
Bump yt-dlp to 2024.03.10 (#109763)

jpbede committed:
Remove deprecated `hass.components` usage in device_sun_light_trigger (#111881)

sdb9696 committed:
Add motion detection enable/disable to ring camera platform (#108789)

* Add motion detection enable/disable to ring camera platform

* Write ha state directly

Co-authored-by: J. Nick Koston <nick@koston.org>

* Parametrize on off state tests

* Add tests for errors on setting motion detection

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

joostlek committed:
Remove entity description mixin in Screenlogic (#112935)

* Remove entity description mixin in Screenlogic

* Fix

jpbede committed:
Remove deprecated `hass.components` usage in service tests (#111883)

StevenLooman committed:
Improve discovering upnp/igd device by always using the SSDP-discovery for the Unique Device Name (#111487)

* Always use the UDN found in the SSDP discovery, instead of the device description

* Ensure existing DeviceEntries are still matched

bieniu committed:
Fix availability for GIOS index sensors (#113021)

* Fix availability for index sensors

* Improve test_availability()

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

cottsay committed:
Fix some handle leaks in rainforest_raven (#113035)

There were leaks when
* The component was shutdown
* There was a timeout during the initial device opening

Additionally, the device was not closed/reopened when there was a
timeout reading regular data.

synesthesiam committed:
Bump intents to 2023.3.12 (#113160)

Bump intents

joostlek committed:
Clean up Abode imports (#113138)

joostlek committed:
Improve lists in integrations [E-F] (#113075)

bdraco committed:
Make august activity update a background task (#112652)

bdraco committed:
Make Bluetooth active coordinator debouncers run tasks in the background (#113129)

bdraco committed:
Ensure HKC debounced availability poll does not block startup (#113130)

bdraco committed:
Avoid multiple executor jobs to add wemo devices (#112484)

Thomas55555 committed:
Bump aioautomower to 2024.3.2 (#113162)

joostlek committed:
Add icon translations to Lametric (#111854)

gjohansson-ST committed:
Add options update listener for Speedtest.Net (#112108)

emontnemery committed:
Add binary_sensor to homeworks (#112418)

* Add binary_sensor to homeworks

* Update tests

Kane610 committed:
Fix Axis unique ID (#112132)

joostlek committed:
Use enum device class in Dexcom (#112423)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

Bre77 committed:
Add icons to Teslemetry (#112546)

mrchi committed:
Bump openwrt-luci-rpc version to 1.1.17 (#112796)

jeeftor committed:
Bump weatherflow4py to 0.1.17 (#112661)

PeteRager committed:
Sonos reduce test time (#111688)

OnFreund committed:
Use friendly name for camera media source (#110882)

bdraco committed:
Avoid some event loop overhead for ESPHome listeners (#113173)

autinerd committed:
Add "language" to strings.json (#112869)

bachya committed:
Move Notion coordinator to its own module (#112756)

Co-authored-by: Cretezy <Cretezy@users.noreply.github.com>

joostlek committed:
Add icon translations to Bond (#111349)

timmo001 committed:
Fix System Bridge media source data URL (#112612)

mattmccormack committed:
Add auto fan mode icon (#110185)

bdraco committed:
Small improvement to test run time (#113175)

TheJulianJES committed:
Remove ZHA IasZone sensor migration (#111893)

bdraco committed:
Add a task name to the config entry retry tasks (#113188)

chamberlain2007 committed:
Code review updates

chamberlain2007 committed:
Code review updates

chamberlain2007 commented:
@joostlek I have moved this to a new PR at #113194
